From 2eeacc219f46edd3bf2c6020b1a726c5073c6f02 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Fri, 20 Dec 2024 11:56:26 +0100
Subject: [PATCH 814/841] fix(st-clock): don't change flexgen for CM33 TDCID on
 STM32MP25

The flexgen configuration (except the flexgen 63 for CA35 bypass) are
managed by cortex M33 when it is TDCID so TF-A BL31 have no access to
associated registers when STM32MP_M33_TDCID is activated.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 22b62febe23c1a17b51a85a5cf7189c9500e1eb7 fix(st-clock): don't change flexgen for CM33 TDCID on STM32MP25]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I3590a0b9d152a0849ae21aa8bd5090b8c9023362
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/442744
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/st/clk/clk-stm32mp2.c | 21 ++++++++++++++-------
 1 file changed, 14 insertions(+), 7 deletions(-)

diff --git a/drivers/st/clk/clk-stm32mp2.c b/drivers/st/clk/clk-stm32mp2.c
index 709de8dfa..dc76239d4 100644
--- a/drivers/st/clk/clk-stm32mp2.c
+++ b/drivers/st/clk/clk-stm32mp2.c
@@ -26,11 +26,14 @@
 #ifdef IMAGE_BL31
 static uint32_t saved_pll_freq1_reg;
 static uint32_t saved_pll_freq2_reg;
+#if !STM32MP_M33_TDCID
 static uint32_t saved_findiv2cfgr;
-static uint32_t saved_findiv5cfgr;
-static uint32_t saved_findiv63cfgr;
 static uint32_t saved_xbar2cfgr;
+
+static uint32_t saved_findiv5cfgr;
 static uint32_t saved_xbar5cfgr;
+#endif /* STM32MP_M33_TDCID */
+static uint32_t saved_findiv63cfgr;
 static uint32_t saved_xbar63cfgr;
 #endif
 
@@ -2862,18 +2865,19 @@ int stm32mp2_pll1_disable(void)
 	saved_pll_freq2_reg = mmio_read_32(pll_freq2_reg);
 
 	/* save FLEXGEN MUX setting */
-	saved_findiv2cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_FINDIV2CFGR);
-	saved_findiv5cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_FINDIV5CFGR);
 	saved_findiv63cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_FINDIV63CFGR);
+	saved_xbar63cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_XBAR63CFGR);
+#if !STM32MP_M33_TDCID
+	saved_findiv2cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_FINDIV2CFGR);
 	saved_xbar2cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_XBAR2CFGR);
+	saved_findiv5cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_FINDIV5CFGR);
 	saved_xbar5cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_XBAR5CFGR);
-	saved_xbar63cfgr = mmio_read_32(stm32mp2_clock_data.base + RCC_XBAR63CFGR);
 
 	/*
 	 * use the clock tree expected by ROM code to avoid a poling timeout
 	 * issue for wake up of Stop2 low power modes, so use HSI for
-	 * flexgen 2 : ck_icn_ddr (ck_icn_m_cpu1)
-	 * flexgen 5 : ck_icn_nic (ck_icn_s_bootrom)
+	 * flexgen 2 : ck_icn_ddr (ck_icn_m_cpu1)            (when CA35TDCID)
+	 * flexgen 5 : ck_icn_nic (ck_icn_s_bootrom)         (when CA35TDCID)
 	 * flexgen 63 : ck_cpu1_ext2f, the CA35 bypass clock
 	 */
 
@@ -2902,6 +2906,7 @@ int stm32mp2_pll1_disable(void)
 			panic();
 		}
 	}
+#endif /* STM32MP_M33_TDCID */
 
 	if ((saved_xbar63cfgr & RCC_XBAR63CFGR_XBAR63SEL_MASK) != XBAR_SRC_HSI) {
 		mmio_clrsetbits_32(stm32mp2_clock_data.base + RCC_XBAR63CFGR,
@@ -2941,6 +2946,7 @@ int stm32mp2_pll1_enable(void)
 		panic();
 	}
 
+#if !STM32MP_M33_TDCID
 	/* restore FLEXGEN MUX setting */
 	if ((saved_xbar2cfgr & RCC_XBAR2CFGR_XBAR2SEL_MASK) != XBAR_SRC_HSI) {
 		mmio_write_32(stm32mp2_clock_data.base + RCC_FINDIV2CFGR, saved_findiv2cfgr);
@@ -2964,6 +2970,7 @@ int stm32mp2_pll1_enable(void)
 		}
 	}
 
+#endif /* STM32MP_M33_TDCID */
 	if ((saved_xbar63cfgr & RCC_XBAR63CFGR_XBAR63SEL_MASK) != XBAR_SRC_HSI) {
 		mmio_write_32(stm32mp2_clock_data.base + RCC_FINDIV63CFGR, saved_findiv63cfgr);
 		if (wait_findivsr(63) != 0) {
-- 
2.47.2

